framework/sessions/__init__.py

```diff
@@ -143,9 +146,36 @@ def create_session(response, data=None):
 # NOTE: This gets attached in website.app.init_app to ensure correct callback
 # order
 def before_request():
+    from framework.auth import authenticate
+    from framework.auth.core import User
+    from framework.auth.cas import CasClient
+
+    # Central Authentication Server Ticket Validation and Authentication
+    ticket = request.args.get('ticket')
+    if ticket:
+        service_url = furl.furl(request.url)
+        service_url.args.pop('ticket')
+
+        cas_resp = CasClient(settings.CAS_SERVER_URL).service_validate(ticket, service_url.url)
+        if cas_resp.authenticated:
+            user = User.load(cas_resp.user)
+            # if we successfully authentication and a verification key is present, invalidate it
+            if user.verification_key:
+                user.verification_key = None
+                user.save()
+            return authenticate(user, cas_resp.attributes['accessToken'], redirect(service_url.url))
+        # Ticket could not be validated, unauthorized.
+        return redirect(service_url.url)
+
+    # Central Authentication Server OAuth Bearer Token
+    authorization = request.headers.get('Authorization')
+    if authorization and authorization.startswith('Bearer '):
+        access_token = authorization[7:]
```

Comment:
This parsing/validation can be made more robust, e.g. if multiple spaces are included or the token is missing, return an error response.

```python
message_short = 'Invalid Bearer token'
parts = request.headers.get('Authorization', '').split()
if parts[0].lower() != 'Bearer:
    message_long = 'Unsupported authorization type'
    raise HTTPError(400, data=dict(
        message_short=message_short,
        message_long=message_long)
    )
elif len(parts) == 1:
    message_long = 'Token missing'
    raise HTTPError(400, data=dict(
        message_short=message_short,
        message_long=message_long)
    )
elif len(parts) > 2:
    message_long = 'Token contains spaces'
    raise HTTPError(400, data=dict(
        message_short=message_short,
        message_long=message_long)
    )
```

Adapted from https://github.com/mattupstate/flask-jwt/blob/master/flask_jwt/__init__.py#L118-L125